#Q1

To calculate the Pearson correlation coefficient between two variables, we need to compute the covariance and the standard deviation of each variable, and then use the formula:

r = cov(X, Y) / (std(X) * std(Y))

where X and Y are the two variables of interest.
Below is example code in python to calculate correlation between time spent studying and Final Exam Score

In [1]:
# Generate synthetic Dataset
import numpy as np
import pandas as pd

# Keep random seed for reproducibility of data
np.random.seed(654)

n = 1000
time_studied = np.random.normal(loc=8, scale=2, size=1000)
score = (time_studied**2)-3*time_studied + 2 + np.random.normal(loc=0, scale=5, size=n)

# create dataframe
df = pd.DataFrame({"time_studied" : time_studied, 
                   "score" : score})
df.head()

time_studied       score
0     11.324031  103.014204
1      5.839362   22.338745
2     10.430491   83.398362
3      7.501036   29.994925
4      9.656534   69.758755

In [2]:
cor = df.corr()
cor

time_studied     score
time_studied      1.000000  0.963761
score             0.963761  1.000000

Positive Correlation found between time studied and Final Score

In [3]:
#Q2

import pandas as pd
dct = {
    'sleep': [8,7,6,5,8,7,6,5,4,7],
    'Job Satisfaction' : [9,6,4,7,8,5,3,6,2,7]
}
df = pd.DataFrame(dct)
df

sleep  Job Satisfaction
0      8                 9
1      7                 6
2      6                 4
3      5                 7
4      8                 8
5      7                 5
6      6                 3
7      5                 6
8      4                 2
9      7                 7

In [4]:
df.rank()

sleep  Job Satisfaction
0    9.5              10.0
1    7.0               5.5
2    4.5               3.0
3    2.5               7.5
4    9.5               9.0
5    7.0               4.0
6    4.5               2.0
7    2.5               5.5
8    1.0               1.0
9    7.0               7.5

In [5]:
df.rank().corr()

sleep  Job Satisfaction
sleep             1.000000          0.660516
Job Satisfaction  0.660516          1.000000

A Spearman's correlation coefficient of 0.66 indicates a moderately strong positive correlation between two variables. The range of Spearman's correlation coefficient is from -1 to +1, where -1 represents a perfectly negative correlation, 0 represents no correlation, and +1 represents a perfectly positive correlation.

In the case of a Spearman's correlation coefficient of 0.66, the two variables being compared have a strong tendency to move together in the same direction. However, the relationship may not be perfect, and there may be some variation or noise in the data.

In [6]:
#Q3

import pandas as pd
import numpy as np

# Set Random seed
np.random.seed(987)

# Generate Sample Data
n = 50
exercise_hours = np.random.normal(loc=5 , scale=1.5, size=n)
BMI = 125/exercise_hours + np.random.normal(loc=0, scale=5, size=n)

# Generate Dataframe
df = pd.DataFrame({'exercise_hours':exercise_hours,
                   'BMI':BMI})
df.head()

exercise_hours        BMI
0        2.471109  55.258468
1        2.793320  39.882961
2        4.832300  32.852014
3        6.554555  11.110522
4        5.991560  20.349710

In [7]:
pearson_corr = df.corr(method='pearson')
pearson_corr


exercise_hours       BMI
exercise_hours        1.000000 -0.850817
BMI                  -0.850817  1.000000

In [8]:
spearman_corr = df.corr(method='spearman')
spearman_corr

exercise_hours       BMI
exercise_hours        1.000000 -0.864298
BMI                  -0.864298  1.000000

Both Spearman's Correlation and Pearson's Correlation show's strong negative Correlation between BMI and exercise time

In [9]:
#Q4

import numpy as np
import pandas as pd
from scipy.stats import pearsonr

# Set the random seed for reproducibility
np.random.seed(123)

# Generate synthetic data for hours of TV watching and physical activity
tv_hours = np.random.normal(loc=3, scale=1, size=50)
physical_activity = 15/tv_hours + np.random.normal(loc=0, scale=1, size=50)

# Showing Dataframe
df = pd.DataFrame({'tv_hours':tv_hours,'physical_activity':physical_activity})
print(df.head())

print('\n===========================================================\n')

# Compute the Pearson correlation coefficient and p-value
corr, pval = pearsonr(tv_hours, physical_activity)

# Print the results
print("Pearson correlation coefficient:", corr)
print("p-value:", pval)

   tv_hours  physical_activity
0  1.914369           6.541393
1  3.997345           2.713702
2  3.282978           6.312734
3  1.493705           9.244079
4  2.421400           6.224447


Pearson correlation coefficient: -0.5795118368116692
p-value: 1.0347687880405448e-05


Moderate Negative Correlation found between tv_hours and phyiscal_activity

In this example, we first import the NumPy library and the pearsonr function from the SciPy library. We then set the random seed for reproducibility and generate synthetic data for the number of hours individuals spend watching television per day and their level of physical activity, using the np.random.normal() function.

Next, we use the pearsonr() function to calculate the Pearson correlation coefficient between the two variables, and store the results in the corr and pval variables. Finally, we print the results to the console.

Note that the pearsonr() function returns two values: the Pearson correlation coefficient and the corresponding p-value. The p-value is a measure of the strength of evidence against the null hypothesis of no correlation, and a p-value less than 0.05 is generally considered to be statistically significant.

In [10]:
#Q5

import pandas as pd
data = {'Age': [25, 42, 37, 19, 31, 28],
        'Soft Drink Preference': ['Coke', 'Pepsi', 'Mountain Dew', 'Coke', 'Pepsi', 'Coke']}
df = pd.DataFrame(data)
df

Age Soft Drink Preference
0   25                  Coke
1   42                 Pepsi
2   37          Mountain Dew
3   19                  Coke
4   31                 Pepsi
5   28                  Coke

In [11]:
# Perform f oneway test
from scipy.stats import f_oneway

coke = df[df['Soft Drink Preference'] == 'Coke']['Age']
pepsi = df[df['Soft Drink Preference'] == 'Pepsi']['Age']
mountain_dew = df[df['Soft Drink Preference'] == 'Mountain Dew']['Age']

f_statistic, p_value = f_oneway(coke, pepsi, mountain_dew)
print('F-statistic:', f_statistic)
print('p-value:', p_value)

F-statistic: 3.524390243902438
p-value: 0.1631217391539759


In [12]:
# Assuming alpha value of 0.05 
alpha = 0.05
if p_value < alpha:
    print('Reject the Null Hypothesis')
    print('Conclusion : Age and Soft Drink Preference are DEPENDENT of each other')
else:
    print('FAIL to reject the Null Hypothesis')
    print('Conclusion : Age and Soft Drink Preference are INDEPENDENT of each other')

FAIL to reject the Null Hypothesis
Conclusion : Age and Soft Drink Preference are INDEPENDENT of each other


In [14]:
#Q6

import numpy as np
import pandas as pd
from scipy.stats import pearsonr

# Generate synthetic data
np.random.seed(1234)  # for reproducibility
sales_calls = np.random.randint(low=50, high=100, size=30)
sales_per_week = np.round(sales_calls * np.random.uniform(low=0.05, high=0.2, size=30), decimals=0)

# Create dataframe for printing
df = pd.DataFrame({'sales_calls':sales_calls,'sales_per_week':sales_per_week})
print(df.head())

print('\n==============================================\n')

# Calculate Pearson correlation coefficient
corr_coef, p_value = pearsonr(sales_calls, sales_per_week)

print("Pearson correlation coefficient: {:.2f}".format(corr_coef))
print("p-value: {:.2f}".format(p_value))

   sales_calls  sales_per_week
0           97            18.0
1           69            14.0
2           88             6.0
3           62             7.0
4           74             9.0


Pearson correlation coefficient: 0.53
p-value: 0.00


Pearson correlation of 0.53 shows moderately positive correlation between sales call per day and sales per week